hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java

```diff
@@ -702,42 +683,41 @@ private void startJoin() {
                 Operation[] postJoinOps = nodeEngine.getPostJoinOperations();
                 boolean createPostJoinOperation = (postJoinOps != null && postJoinOps.length > 0);
-                PostJoinOperation postJoinOp = (createPostJoinOperation ? new PostJoinOperation(postJoinOps) : null);
+                PostJoinOp postJoinOp = (createPostJoinOperation ? new PostJoinOp(postJoinOps) : null);
 
-                clusterService.updateMembers(memberInfos, node.getThisAddress());
+                if (!clusterService.updateMembers(newMembersView, node.getThisAddress(), node.getThisUuid())) {
+                    return;
+                }
 
-                int count = members.size() - 1 + joiningMembers.size();
-                List<Future> calls = new ArrayList<Future>(count);
                 PartitionRuntimeState partitionRuntimeState = partitionService.createPartitionState();
                 for (MemberInfo member : joiningMembers.values()) {
                     long startTime = clusterClock.getClusterStartTime();
-                    Operation finalizeJoinOperation = new FinalizeJoinOperation(member.getUuid(), memberInfos, postJoinOp, time,
-                            clusterService.getClusterId(), startTime,
-                            clusterStateManager.getState(), clusterService.getClusterVersion(), partitionRuntimeState);
-                    calls.add(invokeClusterOperation(finalizeJoinOperation, member.getAddress()));
+                    Operation op = new FinalizeJoinOp(member.getUuid(), newMembersView, postJoinOp, time,
+                            clusterService.getClusterId(), startTime, clusterStateManager.getState(),
+                            clusterService.getClusterVersion(), partitionRuntimeState, true);
+                    op.setCallerUuid(node.getThisUuid());
+                    invokeClusterOp(op, member.getAddress());
                 }
-                for (MemberImpl member : members) {
+                for (MemberImpl member : memberMap.getMembers()) {
                     if (member.localMember() || joiningMembers.containsKey(member.getAddress())) {
                         continue;
                     }
-                    Operation memberInfoUpdateOperation = new MemberInfoUpdateOperation(member.getUuid(), memberInfos, time,
+                    Operation op = new MembersUpdateOp(member.getUuid(), newMembersView, time,
                             partitionRuntimeState, true);
-                    calls.add(invokeClusterOperation(memberInfoUpdateOperation, member.getAddress()));
+                    op.setCallerUuid(node.getThisUuid());
+                    invokeClusterOp(op, member.getAddress());
                 }
 
-                int timeout = Math.min(calls.size() * FINALIZE_JOIN_TIMEOUT_FACTOR, FINALIZE_JOIN_MAX_TIMEOUT);
-                waitWithDeadline(calls, timeout, TimeUnit.SECONDS, whileFinalizeJoinsExceptionHandler);
```

Comment:
Because we reduced periodic membership update duration. Membership publishes can be also triggered via heartbeats etc.